crates/miden-tx/src/executor/notes_checker.rs

```diff
@@ -0,0 +1,83 @@
+use alloc::vec::Vec;
+
+use miden_lib::{
+    account::interface::NoteAccountCompatibility, note::well_known_note::WellKnownNote,
+};
+use miden_objects::{
+    account::AccountId,
+    block::BlockNumber,
+    note::NoteId,
+    transaction::{InputNote, TransactionArgs},
+};
+use winter_maybe_async::{maybe_async, maybe_await};
+
+use super::{ExecutionCheckResult, TransactionExecutor};
+use crate::TransactionExecutorError;
+
+pub struct NotesChecker {
+    account_id: AccountId,
+    notes: Vec<InputNote>,
+}
+
+impl NotesChecker {
+    pub fn new(account_id: AccountId, notes: Vec<InputNote>) -> Self {
+        NotesChecker { account_id, notes }
+    }
+
+    /// Checks whether there are "well known" notes (`P2ID`, `P2IDR` and `SWAP`) in the list of the
+    /// provided input notes. If so, assert that the note inputs are correct.
+    ///
+    /// Returns [NoteAccountCompatibility::No] if at least one note has incorrect inputs.
+    pub fn check_note_inputs(&self) -> (NoteAccountCompatibility, Option<NoteId>) {
```

Comment:
Does this have to be a public API, like is it used in the client or somewhere? If not, then I would make it private.

One way or another, it would be nice to encode this guarantee that if `NoteAccountCompatibility::No` then the `Option` is `Some`, `None` otherwise, because that is not clear without looking at the actual code.

The easiest is to just return `ExecutionCheckResult` directly from this function, but it's not really correct to return `Success` variant from here, because we can't really say that it's a success just yet.

Maybe we can just wrap this in a struct like:

```rust
enum NoteInputsCheck {
    Maybe,
    No { failed_note_id: NoteId }
}
```

and then convert this back into a `NoteAccountCompatibility` in `check_notes_consumability`.